### PR TITLE
[MLIR][NVGPU] Adding `nvgpu.warpgroup.mma` Op for Hopper GPUs

### PR DESCRIPTION
**Pull request by grypp**
This work introduces a new operation called `warpgroup.mma` to the NVGPU dialect of MLIR. The purpose of this operation is to facilitate warpgroup-level matrix multiply and accumulate (WGMMA) operations on Hopper GPUs with sm_90a architecture.

Previously, the `nvvm.wgmma.mma_async` operation was introduced to support warpgroup-level matrix operations in NVVM dialect. This op is used multiple instances of `nvvm.wgmma.mma_async` to achieve the desired shape. The new `nvgpu.warpgroup.mma` operation abstracts this complexity and provides a higher-level interface for performing warpgroup-level matrix operations.

The `nvgpu.warpgroup.mma` does followings:
1) Corresponds multiple `wgmma` instructions.
2) Iterates input matrix descriptors to achieve the desired computation shape. 3) Groups and runs `wgmma` instructions asynchronously, and eventually waits them. This are done by `wgmma.fence.aligned`, `wgmma.commit.group.sync.aligned`, and `wgmma.wait.group.sync.aligned` 4) Results fragmented matrices

Here's an example usage of the `nvgpu.warpgroup.mma` operation:
```
%wgmmaResult, %wgmmaResult2 = nvgpu.warpgroup.mma %descA, %descB, %acc1, %acc2 {transposeB}: 
      !nvgpu.wgmma.descriptor<tensor = memref<128x64xf16, 3>>, 
      !nvgpu.wgmma.descriptor<tensor = memref<64x128xf16, 3>>, 
      !nvgpu.warpgroup.accumulator< fragmented = vector<64x128xf32>>,
      !nvgpu.warpgroup.accumulator< fragmented = vector<64x128xf32>> 
      -> 
      !nvgpu.warpgroup.accumulator< fragmented = vector<64x128xf32>>, 
      !nvgpu.warpgroup.accumulator< fragmented = vector<64x128xf32>>  
```

The op will result following PTX:
```
wgmma.fence.sync.aligned;
wgmma.mma_async.sync.aligned.m64n128k16.f32.f16.f16 {%f1, %f2,    62 more registers}, %descA,     %descB,     p, 1, 1, 0, 1;
wgmma.mma_async.sync.aligned.m64n128k16.f32.f16.f16 {%f1, %f2,    62 more registers}, %descA+2,   %descB+128, p, 1, 1, 0, 1;
wgmma.mma_async.sync.aligned.m64n128k16.f32.f16.f16 {%f1, %f2,    62 more registers}, %descA+4,   %descB+256, p, 1, 1, 0, 1;
wgmma.mma_async.sync.aligned.m64n128k16.f32.f16.f16 {%f1, %f2,    62 more registers}, %descA+8,   %descB+348, p, 1, 1, 0, 1;
wgmma.mma_async.sync.aligned.m64n128k16.f32.f16.f16 {%f500,%f501, 62 more registers}, %descA+512, %descB,     p, 1, 1, 0, 1;
wgmma.mma_async.sync.aligned.m64n128k16.f32.f16.f16 {%f500,%f501, 62 more registers}, %descA+514, %descB+128, p, 1, 1, 0, 1;
wgmma.mma_async.sync.aligned.m64n128k16.f32.f16.f16 {%f500,%f501, 62 more registers}, %descA+516, %descB+256, p, 1, 1, 0, 1;
wgmma.mma_async.sync.aligned.m64n128k16.f32.f16.f16 {%f500,%f501, 62 more registers}, %descA+518, %descB+348, p, 1, 1, 0, 1;
wgmma.commit_group.sync.aligned;
wgmma.wait_group.sync.aligned 1;
```

The Op keeps 
  - first 64 registers (`{%f1, %f2,    62 more registers}`) -> `%acc1` 
  - second 64 registers (`{%f500,%f501, 62 more registers}`)  -> `%acc2`.

### CODE REVIEW

**Comment by joker-eph** [comment]
Typo in the title: `nvgpu.wargroup.mma` -> `nvgpu.warpgroup.mma`

**Comment by grypp** [comment]
> Typo in the title: `nvgpu.wargroup.mma` -> `nvgpu.warpgroup.mma`

Good catch. I somehow wrote wargroup, I put it everywhere :)

File: mlir/include/mlir/Dialect/NVGPU/IR/NVGPU.td [hunk 1]
```diff
@@ -192,6 +192,15 @@ def NVGPU_WarpgroupMatrixDescriptor : NVGPU_Type<"WarpgroupMatrixDescriptor", "w
   let assemblyFormat = "`<` struct(params) `>`";
 }
 
+def NVGPU_WarpgroupResult : NVGPU_Type<"WarpgroupResult", "warpgroup.result", []> {
+  let parameters = (ins "Type":$tensor);
+  let assemblyFormat = "`<` struct(params) `>`";
+  let description = [{
+    It is fragmented result matrix from `nvgpu.warpgroup.mma`.
```
Comment (qcolombet) [inline mlir/include/mlir/Dialect/NVGPU/IR/NVGPU.td:199]: What is "it" in this context?
Maybe something more descriptive:
"This type represents a partial result of the expansion of `nvgpu.warpgroup.mma`?"

Putting an example would be great: Pattern IN -> produced pattern.

File: mlir/include/mlir/Dialect/NVGPU/IR/NVGPU.td [hunk 2]
```diff
@@ -664,5 +673,44 @@ def NVGPU_GenerateGmmaDescriptorOp : NVGPU_Op<"wgmma.generate.descriptor", []> {
...
+    matrix-multiply-and-accumulate (mma) operation that results in 
+    `nvvm.wgmma.mma_async`. 
+    
+    The operands are `descriptorA` and `descriptorB` that are wgmma matrix 
+    descriptors that shows the properties of the matrix in shared memory. The 
+    results are thread-level ownership to the warpgroup-level mma operation 
+    shape. The shape is deduced from the descriptor types and output vector.
+
+    The Op corresponds multiple `nvvm.wgmma.mma_async` operations to complete the 
+    given shape. As the the instruction `nvvm.wgmma.async` is an asyncronous,
```
Comment (qcolombet) [inline mlir/include/mlir/Dialect/NVGPU/IR/NVGPU.td:688]: Typo: the the

Comment (qcolombet) [inline mlir/include/mlir/Dialect/NVGPU/IR/NVGPU.td:688]: typo: is asynchronous

File: mlir/include/mlir/Dialect/NVGPU/IR/NVGPU.td [hunk 3]
```diff
@@ -664,5 +673,44 @@ def NVGPU_GenerateGmmaDescriptorOp : NVGPU_Op<"wgmma.generate.descriptor", []> {
...
+    `wgmma.fence.aligned` and `wgmma.commit.group.sync.aligned`, 
+    `wgmma.wait.group.sync.aligned` Ops.
+
+    Example:
+    ```mlir
+      %res = nvgpu.warpgroup.mma %wgmmaDescA, %wgmmaDescB, %acc: 
+                    !nvgpu.wgmma.descriptor<tensor = memref<128x64xf16, 3>>, 
+                    !nvgpu.wgmma.descriptor<tensor = memref<64x128xf16, 3>>, 
+                    vector<128x128xf32> -> !nvgpu.warpgroup.result<tensor = ...>
+    ```
```
Comment (qcolombet) [inline mlir/include/mlir/Dialect/NVGPU/IR/NVGPU.td:699]: Could you show the expected expansion?

File: mlir/include/mlir/Dialect/NVGPU/IR/NVGPU.td [hunk 4]
```diff
@@ -664,5 +673,44 @@ def NVGPU_GenerateGmmaDescriptorOp : NVGPU_Op<"wgmma.generate.descriptor", []> {
...
+    given shape. As the the instruction `nvvm.wgmma.async` is an asyncronous, 
+    this Op groups the `nvvm.wgmma.async` and surrounds them between 
+    `wgmma.fence.aligned` and `wgmma.commit.group.sync.aligned`, 
+    `wgmma.wait.group.sync.aligned` Ops.
+
+    Example:
+    ```mlir
+      %res = nvgpu.warpgroup.mma %wgmmaDescA, %wgmmaDescB, %acc: 
+                    !nvgpu.wgmma.descriptor<tensor = memref<128x64xf16, 3>>, 
+                    !nvgpu.wgmma.descriptor<tensor = memref<64x128xf16, 3>>,
```
Comment (qcolombet) [inline mlir/include/mlir/Dialect/NVGPU/IR/NVGPU.td:697]: Maybe call out that we only support sizes that are a multiple of `8` and are in `[8; 256]`.
Unless we plan to lift that up?

File: mlir/lib/Conversion/NVGPUToNVVM/NVGPUToNVVM.cpp [hunk 5]
```diff
@@ -1141,6 +1144,164 @@ struct NVGPUTmaCreateDescriptorOpLowering
...
+      wgmmaShapeK = 8;
+    } else if (inputElemType.isF16() || inputElemType.isBF16()) {
+      wgmmaShapeK = 16;
+    } else if (inputElemType.isFloat8E4M3FN() || inputElemType.isFloat8E5M2() ||
+               inputElemType.isInteger(16)) {
+      wgmmaShapeK = 32;
+    } else if (inputElemType.isInteger(1)) {
+      wgmmaShapeK = 256;
+    } else {
+      return failure();
```
Comment (qcolombet) [inline mlir/lib/Conversion/NVGPUToNVVM/NVGPUToNVVM.cpp:1166]: Can this be turned in a `llvm_unreachable`?

Put differently, is it possible to reach that point if the input wgmma passed the verifier?

File: mlir/lib/Conversion/NVGPUToNVVM/NVGPUToNVVM.cpp [hunk 6]
```diff
@@ -1141,6 +1144,164 @@ struct NVGPUTmaCreateDescriptorOpLowering
...
+                            int m, int n, int k, Type resultStructType,
+                            Value inout, Value descriptorA,
+                            Value descriptorB) const {
+    TypeRange resultTypes = {resultStructType};
+    auto shape = NVVM::MMAShapeAttr::get(ctx, m, n, k);
+    auto scaleOut = NVVM::WGMMAScaleOutAttr::get(ctx, NVVM::WGMMAScaleOut::one);
+    auto scaleIn = NVVM::WGMMAScaleInAttr::get(ctx, NVVM::WGMMAScaleIn::one);
+    auto layoutA = NVVM::MMALayoutAttr::get(ctx, NVVM::MMALayout::row);
+    auto layoutB = NVVM::MMALayoutAttr::get(ctx, NVVM::MMALayout::col);
+    // todo input type
```
Comment (qcolombet) [inline mlir/lib/Conversion/NVGPUToNVVM/NVGPUToNVVM.cpp:1185]: What does this todo entail?

**Review by qcolombet — APPROVED** [review]
Looks mostly good to me.
Highlighted a few places where comments would make things easier to understand.

File: mlir/include/mlir/Dialect/NVGPU/IR/NVGPU.td (hunk 4 above)
Comment (grypp) [reply mlir/include/mlir/Dialect/NVGPU/IR/NVGPU.td]: Not sure I understood this comment

**Pull request merged**